snapshot_creator.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2021 the v8go contributors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package v8go
+
+// #include <stdlib.h>
+// #include "v8go.h"
+import "C"
+import (
+	"errors"
+	"unsafe"
+)
+
+type FunctionCodeHandling int
+
+//  Clear - does not keeps any compiled data prior to serialization/deserialization/verify pass
+//  Keep - keeps any compiled data prior to serialization/deserialization/verify pass
+const (
+	FunctionCodeHandlingClear FunctionCodeHandling = iota
+	FunctionCodeHandlingKeep
+)
+
+// StartupData stores the snapshot blob data
+type StartupData struct {
+	data     []byte
+	raw_size C.int
+}
+
+// SnapshotCreator allows creating snapshot.
+type SnapshotCreator struct {
+	ptr                 C.SnapshotCreatorPtr
+	iso                 *Isolate
+	defaultContextAdded bool
+}
+
+// NewSnapshotCreator creates a new snapshot creator.
+func NewSnapshotCreator() *SnapshotCreator {
+	v8once.Do(func() {
+		C.Init()
+	})
+
+	rtn := C.NewSnapshotCreator()
+
+	return &SnapshotCreator{
+		ptr:                 rtn.creator,
+		iso:                 &Isolate{ptr: rtn.iso},
+		defaultContextAdded: false,
+	}
+}
+
+// GetIsolate returns the Isolate associated with the SnapshotCreator.
+// This Isolate must be use to create the contexts that later will be use to create the snapshot blob.
+func (s *SnapshotCreator) GetIsolate() (*Isolate, error) {
+	if s.ptr == nil {
+		return nil, errors.New("v8go: Cannot get Isolate after creating the blob")
```

Comment:
Might be worth panicking here trying to call a function on a snapshot creator thats likely been disposed of (either through `Dispose` or from calling `Create` to make the blob)

Reply:
It seems that the pattern among other functions https://github.com/rogchap/v8go/pull/118#discussion_r619480410

Should we panic in all of the instances. It seems that it would make working with the Snapshot creator API much easier. 

```golang
snapshotCreator := v8.NewSnapshotCreator()
snapshotCreatorIso := snapshotCreator.GetIsolate()

snapshotCreatorCtx := v8.NewContext(snapshotCreatorIso)
defer snapshotCreatorCtx.Close()

snapshotCreatorCtx.RunScript(`const add = (a, b) => a + b`, "add.js")
snapshotCreatorCtx.RunScript(`function run() { return add(3, 4); }`, "main.js")
snapshotCreator.SetDefaultContext(snapshotCreatorCtx)
data := snapshotCreator.Create(v8.FunctionCodeHandlingClear)

iso := v8.NewIsolate(v8.WithStartupData(data))
defer iso.Dispose()

ctx, err := v8.NewContextFromSnapshot(iso, index)
defer ctx.Close()
```

The only one that I'm not sure 100% we should panic is the one that errors when calling `Create` before adding a default context since this is something a developer can fix. 

What do you think @genevieve?

Reply:
Lets stick the error